uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js

```diff
@@ -0,0 +1,516 @@
+       */
+      $ctrl.changeInNoOfDays = function () {
+        reset();
+        //reinitialize opening balance
+        $ctrl.balance.opening = $ctrl.selectedAbsenceType.remainder;
+      };
+
+      /**
+       * Whenever the absence type changes, update the balance opening.
+       * Also the balance change needs to be recalculated, if the `from` and `to`
+       * dates have been already selected
+       */
+      $ctrl.onAbsenceTypeChange = function () {
+        $ctrl.leaveRequest.type_id = $ctrl.selectedAbsenceType.id;
+        // get the `balance` of the newly selected absence type
+        $ctrl.balance.opening = $ctrl.selectedAbsenceType.remainder;
+
+        if ($ctrl.leaveRequest.from_date && $ctrl.leaveRequest.to_date) {
+          $ctrl.loading.calculateBalanceChange = true;
+          $ctrl.calculateBalanceChange().then(function () {
+            $ctrl.loading.calculateBalanceChange = false;
+          });
+        }
+      };
+
+      /**
+       * This should be called whenever a date has been changed
+       *
+       * First it syncs `from` and `to` date, if it's in 'single day' mode
+       * Then, if all the dates are there, it gets the balance change
+       *
+       */
+      $ctrl.onDateChange = function (date, isFrom) {
+        if ($ctrl.uiOptions.multipleDays) {
+          if ($ctrl.uiOptions.toDate && $ctrl.uiOptions.fromDate) {
+            $ctrl.uiOptions.showBalance = true;
+          }
+        } else {
+          if ($ctrl.uiOptions.fromDate) {
+            $ctrl.uiOptions.showBalance = true;
+          }
+        }
+
+        $ctrl.filterLeaveRequestDayTypes(date, !!isFrom)
+          .then(function () {
+            if (isFrom) {
+              $ctrl.leaveRequest.from_date = convertDateFormatToServer($ctrl.uiOptions.fromDate);
+              $ctrl.loading.fromDayTypes = false;
+            } else {
+              $ctrl.leaveRequest.to_date = convertDateFormatToServer($ctrl.uiOptions.toDate);
+              $ctrl.loading.toDayTypes = false;
+            }
+
+            if (!$ctrl.uiOptions.multipleDays) {
+              $ctrl.uiOptions.toDate = $ctrl.uiOptions.fromDate;
+              $ctrl.uiOptions.selectedToType = $ctrl.uiOptions.selectedFromType;
+              $ctrl.leaveRequest.to_date = $ctrl.leaveRequest.from_date;
+              $ctrl.leaveRequest.to_date_type = $ctrl.leaveRequest.from_date_type;
+            }
+
+            if (!$ctrl.uiOptions.showBalance) {
+              return
+            }
+
+            $ctrl.loading.calculateBalanceChange = true;
+            if ($ctrl.leaveRequest.from_date && $ctrl.leaveRequest.to_date) {
+              $ctrl.calculateBalanceChange().then(function () {
+                $ctrl.loading.dateChangesBalance = false;
+              });
+            }
+          });
+      };
+
+      /**
+       * Calculate change in balance, it updates local balance variables
+       *
+       * @return {Promise}
+       */
+      $ctrl.calculateBalanceChange = function () {
```

Comment:
Another example of just *too much* going on in a given method. Methods should be single-purpose and as short as possible. Please refactor this